### PR TITLE
Kafka-mesh filter

### PR DESCRIPTION
**Pull request by adamkotwasinski**
Commit message: Kafka-mesh filter
Additional Description: 
Kafka mesh filter.
Can be put between native Kafka client and intercept produce requests to forward them to upstream Kafka clusters as per its fowarding rules. Right now there's only one type of rule - prefix by topic name (so e.g. we can send messages to topics `a*` and have them in cluster 1, while others go to cluster 2). Only produce requests for now. Support for consumers would be more complex or require more limitations (as Kafka consumers keep state, so it's not like Redis - or actually it's like _Redis pub/sub_).
Doesn't work on Windows because of librdkafka compilation issues (and I don't have a Windows machine).
Risk Level: Low
Testing: Unit tests + integration test (2 kafka clusters are started + proxy in between + producer) + [manual testing](https://github.com/adamkotwasinski/envoy-kafka-mesh-test/tree/main/src/test/java/envoy).
Docs Changes: New .rst & .proto files, as we have a new filter.
Release Notes: N/A

overview-incoming-request

Without getting into too much details:
1) Kafka producers (Java, Python, etc.) send records to Envoy (the filter contains enough code to establish communication, first it needs to advertise itself it's a Kafka server through api-versions + metadata requests).
2) Kafka-mesh filter instance receives the records, parses them, forwards them to rules which decide where the records should go.
3) The records are forwarded to producers dedicated for each of clusters.
4) librdkafka producers send the records where they should have gone (librdkafka has its own threads and we use `dispatcher.post` to have the delivery confirmation back to normal worker thread).

----

Other PRs for kafka-mesh work (merged independently):
https://github.com/envoyproxy/envoy/pull/17202
https://github.com/envoyproxy/envoy/pull/17374
https://github.com/envoyproxy/envoy/pull/17475
https://github.com/envoyproxy/envoy/pull/17597
https://github.com/envoyproxy/envoy/pull/17783
https://github.com/envoyproxy/envoy/pull/17818
https://github.com/envoyproxy/envoy/pull/17839
https://github.com/envoyproxy/envoy/pull/17887

### CODE REVIEW

**Comment by adamkotwasinski** [comment]
test

**Comment by adamkotwasinski** [comment]
still alive just busy with other stuff

**Comment by adamkotwasinski** [comment]
I'll be back to working on this in ~2 weeks

**Comment by antoniovicente** [comment]
Best of luck getting this closer to a reviewable state.  I recommend you consider breaking this change into smaller pieces that can be merged independently.  Of course, each piece should be mostly independent and have tests to validate the functionality it adds.

**Comment by adamkotwasinski** [comment]
All right, it "compiles" but is not yet ready. What's left:
* integration tests for clients with linger.ms > 0
* docs: only magic = 2 (check versions)
* docs: python producer config
* docs: reordering / de-bunching semantics
* remove KafkaProducerWrapperImpl?
* switch in splitter causing coverage issues (!)
* rename splitter to something else
* missing unit tests in serialization (feeding byte-by-byte)

**Comment by adamkotwasinski** [comment]
reopen

**Comment by adamkotwasinski** [comment]
/wait - this is not to be reviewed :)

**Comment by adamkotwasinski** [comment]
don't look at this PR - it's just to verify builds etc.

**Comment by adamkotwasinski** [comment]
I don't think we'll need this PR that much anymore, as with move to contrib, our tests are not longer executed in CI. Separate PRs will follow.

**Comment by mattklein123** [comment]
> I don't think we'll need this PR that much anymore, as with move to contrib, our tests are not longer executed in CI. Separate PRs will follow.

FYI your tests are run in CI, just only under the release build.

**Comment by adamkotwasinski** [comment]
docs & some other minor stuff needs some tuning, so not yet fully ready

**Comment by adamkotwasinski** [comment]
Ready for review - also ran some Java tests.

File: docs/root/configuration/listeners/network_filters/kafka_mesh_filter.rst [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+Kafka Mesh filter
+===================
+
+The Apache Kafka mesh filter provides a facade for `Apache Kafka <https://kafka.apache.org/>`_
+producers. Produce requests sent to this filter insance can be forwarded to one of multiple
+clusters, depending on configured forwarding rules. Corresponding message versions from
+Kafka 2.4.0 are supported.
+
+.. attention::
```
Comment (mattklein123) [inline docs/root/configuration/listeners/network_filters/kafka_mesh_filter.rst:11]: Can you also add an attention/warning here that this is only available on contrib builds with a link? See what we do for squash and some of the other contrib filters and please unify as needed, thanks.

File: source/extensions/filters/network/well_known_names.h [hunk 2]
```diff
@@ -47,6 +47,8 @@ class NetworkFilterNameValues {
   const std::string ExtAuthorization = "envoy.filters.network.ext_authz";
   // Kafka Broker filter
   const std::string KafkaBroker = "envoy.filters.network.kafka_broker";
+  // Kafka Mesh filter
+  const std::string KafkaMesh = "envoy.filters.network.kafka_mesh";
```
Comment (mattklein123) [inline source/extensions/filters/network/well_known_names.h:51]: we are trying to get rid of this file. Can you just inline this?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Cool, thanks. A few small comments.

/wait

File: docs/root/configuration/listeners/network_filters/kafka_mesh_filter.rst (hunk 1 above)
Comment (adamkotwasinski) [reply docs/root/configuration/listeners/network_filters/kafka_mesh_filter.rst:18]: done - comments and missing link added; same for kafka-broker

File: source/extensions/filters/network/well_known_names.h (hunk 2 above)
Comment (adamkotwasinski) [reply source/extensions/filters/network/well_known_names.h]: done

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**